plugins/image-prioritizer/class-image-prioritizer-video-tag-visitor.php

```diff
@@ -35,15 +35,86 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 			return false;
 		}
 
-		// Skip empty poster attributes and data: URLs.
-		$poster = trim( (string) $processor->get_attribute( 'poster' ) );
+		// TODO: If $context->url_metric_group_collection->get_element_max_intersection_ratio( $xpath ) is 0.0, then the video is not in any initial viewport and the VIDEO tag could get the preload=none attribute added.
+
+		$reduced_poster_size = $this->reduce_poster_image_size( $context );
+		$preload_poster      = $this->preload_poster_image( $context );
+
+		return $reduced_poster_size || $preload_poster;
+	}
+	/**
+	 * Gets the poster from the current VIDEO element.
+	 *
+	 * Skips empty poster attributes and data: URLs.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 * @return non-empty-string|null Poster or null if not defined or is a data: URL.
+	 */
+	private function get_poster( OD_Tag_Visitor_Context $context ): ?string {
+		$poster = trim( (string) $context->processor->get_attribute( 'poster' ) );
 		if ( '' === $poster || $this->is_data_url( $poster ) ) {
-			return false;
+			return null;
 		}
+		return $poster;
+	}
+
+	/**
+	 * Reduces poster image size by choosing one that fits the maximum video size more closely.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context, with the cursor currently at a VIDEO tag.
+	 * @return bool Whether the tag should be tracked in URL metrics.
+	 */
+	private function reduce_poster_image_size( OD_Tag_Visitor_Context $context ): bool {
+		$processor = $context->processor;
 
 		$xpath = $processor->get_xpath();
 
-		// TODO: If $context->url_metric_group_collection->get_element_max_intersection_ratio( $xpath ) is 0.0, then the video is not in any initial viewport and the VIDEO tag could get the preload=none attribute added.
+		$max_element_width = 0;
+
+		$denormalized_elements = $context->url_metric_group_collection->get_all_denormalized_elements()[ $xpath ] ?? array();
+
+		foreach ( $denormalized_elements as list( , , $element ) ) {
+			$max_element_width = max( $max_element_width, $element['boundingClientRect']['width'] ?? 0 );
+		}
+
+		$poster = $this->get_poster( $context );
```

Comment:
What about passing the `$poster` in as an argument to `reduce_poster_image_size()` and `preload_poster_image()` as shown in https://github.com/WordPress/performance/pull/1595#discussion_r1803433939?

Reply:
Ah, sorry, missed that second diff.

Doesn't make a big difference to me, I can add it.

But the early return in `__invoke` will need to change a little bit in #1596.